[nnc] Update cuda codegen to use llvm for thread and block extent computations

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #72077
* #72036
* #72038
* __->__ #72040
* #72032
* #71651
* #71650
* #71173

Differential Revision: [D33865041](https://our.internmc.facebook.com/intern/diff/D33865041)

Commit:
[nnc] Update cuda codegen to use llvm for thread and block extent computations

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33865041).

Commit:
Update on "[nnc] Update cuda codegen to use llvm for thread and block extent computations"


Differential Revision: [D33865041](https://our.internmc.facebook.com/intern/diff/D33865041)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33865041).

navahgar commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -1024,6 +1024,46 @@ void CudaCodeGen::Initialize() {
+  for (const size_t i : c10::irange(buffer_args.size())) {
+    if (vars_in_extents.count(buffer_args[i].var())) {
+      extents_buffer_args.push_back(buffer_args[i]);
+      arg_pos_in_extents_.push_back(true);
```
Add `.reserve` for this.

eellison replied:
this is in initialize, not hot path, so perf doesn't matter as much here. but i can add

navahgar commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -1024,6 +1024,46 @@ void CudaCodeGen::Initialize() {
 
+  // Build an LLVM based eval expression for the extents
+  block_extents_eval_.reserve(block_extents.size());
+  std::vector<BufferArg> extents_buffer_args;
```
Can you add the following comment here?

```
We need to extract the args that are used in the thread and block extents 
from bufferArgs and only use those for the `ExprEval` below. Without this, 
bufferArgs might contain arbitrary types that are not handled by LLVM and 
hence would result in an error.
```

navahgar reviewed (approved):
I thought you were going to use Mike's approach here. Something changed with that?

Commit:
Update on "[nnc] Update cuda codegen to use llvm for thread and block extent computations"


Differential Revision: [D33865041](https://our.internmc.facebook.com/intern/diff/D33865041)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33865041).
